tests/unit/s2n_certificate_test.c

```diff
@@ -754,7 +755,184 @@ int main(int argc, char **argv)
+            EXPECT_SUCCESS(s2n_connection_set_blinding(server, S2N_SELF_SERVICE_BLINDING));
+            EXPECT_SUCCESS(s2n_connection_set_config(server, server_config));
+
+            DEFER_CLEANUP(struct s2n_test_io_pair io_pair = { 0 }, s2n_io_pair_close);
+            EXPECT_SUCCESS(s2n_io_pair_init_non_blocking(&io_pair));
+            EXPECT_SUCCESS(s2n_connections_set_io_pair(client, server, &io_pair));
+
+            EXPECT_FAILURE_WITH_ERRNO(s2n_negotiate_test_server_and_client(server, client),
+                    S2N_ERR_CERT_UNTRUSTED);
+            /* Both the client and server could produce S2N_ERR_CERT_UNTRUSTED.
+             * Verify that only the server encountered an error and therefore only
+             * the server closed the connection.
+             */
+            EXPECT_TRUE(s2n_connection_check_io_status(server, S2N_IO_CLOSED));
+            EXPECT_FALSE(s2n_connection_check_io_status(client, S2N_IO_CLOSED));
+
+            EXPECT_NOT_EQUAL(server->handshake_params.client_cert_chain.size, 0);
+            EXPECT_NOT_NULL(server->handshake_params.client_cert_chain.data);
+        };
+
+        /* Test: Malformed TLS1.2 certificate chain is still available */
+        {
+            DEFER_CLEANUP(struct s2n_stuffer input, s2n_stuffer_free);
+            EXPECT_SUCCESS(s2n_stuffer_growable_alloc(&input, 0));
+
+            struct s2n_stuffer_reservation total_size = { 0 };
+            const uint32_t cert_size = 24;
+            EXPECT_SUCCESS(s2n_stuffer_reserve_uint24(&input, &total_size));
+            EXPECT_SUCCESS(s2n_stuffer_write_uint24(&input, cert_size));
+            EXPECT_SUCCESS(s2n_stuffer_skip_write(&input, cert_size));
+            EXPECT_SUCCESS(s2n_stuffer_write_uint24(&input, cert_size));
+            EXPECT_SUCCESS(s2n_stuffer_skip_write(&input, cert_size));
+            EXPECT_SUCCESS(s2n_stuffer_write_vector_size(&total_size));
+
+            /* Sanity check: test cert chain is well-formed (but invalid) */
+            {
+                DEFER_CLEANUP(struct s2n_connection *server = s2n_connection_new(S2N_SERVER),
+                        s2n_connection_ptr_free);
+                server->actual_protocol_version = S2N_TLS12;
+                EXPECT_SUCCESS(s2n_stuffer_copy(&input, &server->handshake.io,
+                        s2n_stuffer_data_available(&input)));
+
+                EXPECT_FAILURE_WITH_ERRNO(s2n_client_cert_recv(server), S2N_ERR_CERT_INVALID);
+                EXPECT_NOT_EQUAL(server->handshake_params.client_cert_chain.size, 0);
+                EXPECT_NOT_NULL(server->handshake_params.client_cert_chain.data);
+            };
+
+            /* Modify last cert to be malformed. Keep correct total size */
+            EXPECT_SUCCESS(s2n_stuffer_reread(&input));
+            EXPECT_SUCCESS(s2n_stuffer_wipe_n(&input, 1));
+            EXPECT_SUCCESS(s2n_stuffer_write_vector_size(&total_size));
+
+            /* Data available even if malformed: we don't parse pre-TLS1.3 chains
```

Comment:
What's the difference between this test and the previous one?

Reply:
If you mean what's the difference between "malformed" and "fails validation": Malformed input can't even be parsed out into a chain that we can validate. The Certificate message is malformed.

If by previous one you mean the case labeled "sanity check", that's not really a test. It's just to make sure that the input is valid before we intentionally modify it to make it malformed. That way it's malformed in only the way we expect.

Are either of those what you meant?

Reply:
It seems like your sanity check and the real test are actually two separate tests, one where you're proving that invalid certificate chains are stored, and one where you're proving that malformed certificate chains are stored. The "success criteria"(lines 856-858) are the same in both.

Reply:
I reorganized it a bit and added more comments -- see if that helps?